File: lib/audiotag/id3v1.ex
defmodule AudioTag.ID3v1 do
  @behaviour AudioTag.FrameParser

  defmodule Frame do
    defstruct [:title, :artist, :album, :year, :comment, :genre]
  end

  def matches?(reader) do
    case AudioTag.FileReader.peek(reader, 3) do
      {:ok, <<"TAG">>, reader} ->
        {true, reader}
      {:eof, reader} ->
        {false, reader}
    end
  end

  def parse(reader) do
    case AudioTag.FileReader.read(reader, 128) do
      {:ok, data, reader} ->
        {:ok, parse_frame(data), reader}
      {:eof, reader} ->
        {:error, :eof, reader}
    end
  end

  defp parse_frame(<<"TAG",
                   title::bytes-size(30),
                   artist::bytes-size(30),
                   album::bytes-size(30),
                   year::bytes-size(4),
                   comment::bytes-size(30),
                   genre::bytes-size(1)>>) do
    %Frame{
      title: strip_field(title),
      artist: strip_field(artist),
      album: strip_field(album),
      year: year,
      comment: strip_field(comment),
      genre: genre
    }
  end

  def strip_field(data) do
    :binary.split(data, <<0>>) |> List.first
  end
end

File: audio/parsers/mp3/id3v1.go
package mp3

type ID3v1Tag struct {
	Raw []byte

	Title   string
	Artist  string
	Album   string
	Year    string
	Comment string
	Genre   int
}

func IsID3v1Frame(buf []byte) bool {
	if len(buf) < 3 {
		return false
	}

	return buf[0] == 'T' && buf[1] == 'A' && buf[2] == 'G'
}

func readID3v1String(buf []byte) string {
	for i := 0; i < len(buf); i++ {
		if buf[i] == 0 {
			return string(buf[:i])
		}
	}

	return string(buf)
}

func (f *ID3v1Tag) Parse() {
	f.Title = readID3v1String(f.Raw[3:33])
	f.Artist = readID3v1String(f.Raw[33:63])
	f.Album = readID3v1String(f.Raw[63:93])
	f.Year = string(f.Raw[93:97])
	f.Comment = readID3v1String(f.Raw[97:127])
	f.Genre = int(f.Raw[127])
}
